### PR TITLE
Some python 2 cleanup

### PR DESCRIPTION
**Pull request by lucas-wilkins**
Removed some of the python 2 legacy. Mostly superficial, added some type hints and renamed a few things.

I have changed some more fundamental things in the loaders, this is most likely place for bugs.

### CODE REVIEW

File: src/sas/sascalc/dataloader/loader.py [hunk 1]
```diff
@@ -281,63 +283,62 @@ def _identify_plugin(self, module):
...
-        # Remove duplicates if they exist
-        if len(writers) != len(set(writers)):
-            result = []
-            for L in writers:
-                if L not in result:
-                    result.append(L)
-            writers = L
-        # Raise an error if there are no matching extensions
+
+        writers = [writer for ext in extlist for writer in self.writers[ext]]
```
Comment (rprospero) [inline src/sas/sascalc/dataloader/loader.py:331]: I like the use of list comprehensions for these sorts of problems.

**Review by rprospero — CHANGES_REQUESTED** [review]
Mostly looks pretty good.  I've got a couple of stylistic comments, but I also wouldn't mind someone else who's worked on this part of the code taking a look before I merge this.

File: src/sas/sascalc/data_util/registry.py [hunk 2]
```diff
@@ -59,62 +62,56 @@ class ExtensionRegistry(object):
...
         """
+
         # Find matching lower-case extensions
         path_lower = path.lower()
-        extlist = [ext for ext in self.extensions() if path_lower.endswith(ext)]
-        # Sort matching extensions by decreasing order of length
-        extlist.sort(key=len)
+        extensions = [ext for ext in self.extensions() if path_lower.endswith(ext)]
+
+        # Sort matching extensions by decreasing order of length # TODO: Again, why???
```
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/registry.py:105]: OK, probably a good thing I left it in. It should probably be handled in a better way in general, or at least there should be a comment that explains why its being done.

**Comment by lucas-wilkins** [comment]
@rprospero if there is someone you think should review it, can you ask for a review from them. I'll also flag this PR for the meeting today.

File: src/sas/sascalc/data_util/registry.py [hunk 3]
```diff
@@ -59,62 +62,56 @@ class ExtensionRegistry(object):
...
-                    result.append(L)
-            loaders = result
-        return loaders
-
-    def load(self, path, format=None):
+        loaders = [loader for ext in extensions for loader in self.readers[ext]]
+
+        # Remove duplicates and return
+        return list(set(loaders))
+
```
Comment (rozyczko) [inline src/sas/sascalc/data_util/registry.py:113]: Won't this modify the ordering of the original list? It has just been sorted in line 106

File: src/sas/sascalc/data_util/registry.py [hunk 4]
```diff
@@ -59,62 +62,56 @@ class ExtensionRegistry(object):
...
-        return ext in self.loaders
+    def __contains__(self, ext: str):
+        return ext in self.readers
 
     def formats(self):
         """
         Return a sorted list of the registered formats.
         """
-        names = [a for a in self.loaders.keys() if not a.startswith('.')]
+        names = [a for a in self.readers.keys() if not a.startswith('.')] # What is this doing?
```
Comment (krzywon) [inline src/sas/sascalc/data_util/registry.py:81]: This is filtering out built-in readers from user-defined ones. It's possible for a user to associate a reader with another extension and user-defined associations should have no leading `.`. This is not currently exposed in the GUI.

File: src/sas/sascalc/data_util/registry.py (hunk 3 above)
Comment (krzywon) [reply src/sas/sascalc/data_util/registry.py:113]: Good catch. Moving the sort to the return statement would fix this.

File: src/sas/sascalc/data_util/registry.py [hunk 5]
```diff
@@ -59,62 +62,56 @@ class ExtensionRegistry(object):
...
         """
         Return a sorted list of registered extensions.
         """
-        exts = [a for a in self.loaders.keys() if a.startswith('.')]
+        exts = [a for a in self.readers.keys() if a.startswith('.')]
         exts.sort()
         return exts
 
-    def lookup(self, path):
+    def lookup(self, path: str):
```
Comment (krzywon) [inline src/sas/sascalc/data_util/registry.py:93]: Return type?

File: src/sas/sascalc/data_util/registry.py (hunk 3 above)
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/registry.py:113]: yeah, true.

File: src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py [hunk 6]
```diff
@@ -657,16 +654,24 @@ def _initialize_new_data_set(self, value=None):
         self.current_datainfo.filename = self.raw_data.filename
 
     @staticmethod
-    def as_list_or_array(iterable):
+    def as_list_or_array(data: Iterable):
```
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py:657]: yeah, will sort that. Think it should be something like Union[some things that are checked] -> Iterable

File: src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py [hunk 7]
```diff
@@ -657,16 +654,24 @@ def _initialize_new_data_set(self, value=None):
...
-        :param iterable:
+        :param data:
         :return:
         """
-        if not (isinstance(iterable, np.ndarray) or isinstance(iterable, list)):
-            iterable = iterable.split(",") if isinstance(iterable, basestring)\
-                else [iterable]
-        return iterable
+
+        if isinstance(data, np.ndarray):
```
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py:664]: yeah

File: src/sas/sascalc/data_util/registry.py (hunk 5 above)
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/registry.py:93]: Can do, though I tend to follow the convention that return types need not be specified unless they differ from what would be inferred.

File: src/sas/sascalc/data_util/registry.py [hunk 8]
```diff
@@ -59,62 +62,56 @@ class ExtensionRegistry(object):
...
-    def __init__(self, **kw):
-        self.loaders = {}
+    def __init__(self):
+        self.readers = defaultdict(list)
 
     def __setitem__(self, ext, loader):
-        if ext not in self.loaders:
-            self.loaders[ext] = []
-        self.loaders[ext].insert(0, loader)
+        self.readers[ext].insert(0, loader) # TODO: Why insert at zero, not just append?
```
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/registry.py:69]: OK

File: src/sas/sascalc/data_util/registry.py (hunk 5 above)
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/registry.py]: Changed

File: src/sas/sascalc/data_util/registry.py (hunk 4 above)
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/registry.py]: Thanks

File: src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py:664]: done

File: src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/sascalc/dataloader/readers/cansas_reader_HDF5.py:657]: actually Any -> Union[list, np.ndarray] - done

File: src/sas/sascalc/data_util/registry.py (hunk 3 above)
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/registry.py:113]: fixed

File: src/sas/sascalc/data_util/registry.py [hunk 9]
```diff
@@ -59,62 +62,56 @@ class ExtensionRegistry(object):
         registry.load('hello.cx',format='cx3') ->
             return cx3('hello.cx')
     """
-    def __init__(self, **kw):
-        self.loaders = {}
+    def __init__(self):
+        self.readers = defaultdict(list)
```
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/registry.py:66]: Lets discuss names tomorrow

**Review by rprospero — APPROVED** [review]

**Pull request merged**